File: go/runtime/environment.go
package runtime

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"github.com/ter-sdk/ter-go/schema"
)

// Environment provides type-safe environment access
type Environment struct {
	schema   *schema.Schema
	resolver *Resolver
}

// NewEnvironment creates a new environment
func NewEnvironment(s *schema.Schema, r *Resolver) *Environment {
	return &Environment{
		schema:   s,
		resolver: r,
	}
}

// Get returns a raw value
func (e *Environment) Get(name string, defaultVal interface{}) interface{} {
	resolved := e.resolver.Resolve(name)
	if resolved != nil {
		return resolved.Value
	}
	return defaultVal
}

// GetString returns a string value
func (e *Environment) GetString(name string) (string, error) {
	val := e.Get(name, nil)
	if val == nil {
		return "", fmt.Errorf("%s is required", name)
	}
	return fmt.Sprintf("%v", val), nil
}

// GetInt returns an integer value
func (e *Environment) GetInt(name string) (int, error) {
	val := e.Get(name, nil)
	if val == nil {
		return 0, fmt.Errorf("%s is required", name)
	}

	switch v := val.(type) {
	case int:
		return v, nil
	case int64:
		return int(v), nil
	case float64:
		return int(v), nil
	case string:
		n, err := strconv.Atoi(v)
		if err != nil {
			return 0, fmt.Errorf("cannot convert '%v' to int", val)
		}
		return n, nil
	default:
		return 0, fmt.Errorf("cannot convert '%v' to int", val)
	}
}

// GetFloat returns a float value
func (e *Environment) GetFloat(name string) (float64, error) {
	val := e.Get(name, nil)
	if val == nil {
		return 0, fmt.Errorf("%s is required", name)
	}

	switch v := val.(type) {
	case float64:
		return v, nil
	case int:
		return float64(v), nil
	case string:
		f, err := strconv.ParseFloat(v, 64)
		if err != nil {
			return 0, fmt.Errorf("cannot convert '%v' to float", val)
		}
		return f, nil
	default:
		return 0, fmt.Errorf("cannot convert '%v' to float", val)
	}
}

// GetBool returns a boolean value
func (e *Environment) GetBool(name string) (bool, error) {
	val := e.Get(name, nil)
	if val == nil {
		return false, fmt.Errorf("%s is required", name)
	}

	switch v := val.(type) {
	case bool:
		return v, nil
	case string:
		lower := strings.ToLower(v)
		if lower == "true" || lower == "1" || lower == "yes" || lower == "on" {
			return true, nil
		}
		if lower == "false" || lower == "0" || lower == "no" || lower == "off" {
			return false, nil
		}
		return false, fmt.Errorf("'%v' is not a valid boolean", val)
	default:
		return false, fmt.Errorf("cannot convert '%v' to bool", val)
	}
}

// GetJSON returns a JSON-decoded value
func (e *Environment) GetJSON(name string) (interface{}, error) {
	val := e.Get(name, nil)
	if val == nil {
		return nil, fmt.Errorf("%s is required", name)
	}

	switch v := val.(type) {
	case map[string]interface{}, []interface{}:
		return v, nil
	case string:
		var result interface{}
		if err := json.Unmarshal([]byte(v), &result); err != nil {
			return nil, fmt.Errorf("invalid JSON: %w", err)
		}
		return result, nil
	default:
		return nil, fmt.Errorf("cannot decode '%v' as JSON", val)
	}
}

// Validate validates all variables
func (e *Environment) Validate() (map[string]interface{}, error) {
	values := make(map[string]interface{})
	for name := range e.schema.GetVariables() {
		resolved := e.resolver.Resolve(name)
		if resolved != nil {
			values[name] = resolved.Value
		}
	}
	return e.schema.Validate(values)
}

// ToMap converts to map representation
func (e *Environment) ToMap() map[string]interface{} {
	m := make(map[string]interface{})
	for name := range e.schema.GetVariables() {
		val := e.Get(name, nil)
		if val != nil {
			m[name] = val
		}
	}
	return m
}

// GetMetadata returns metadata about a value
func (e *Environment) GetMetadata(name string) map[string]interface{} {
	resolved := e.resolver.Resolve(name)
	if resolved != nil {
		return map[string]interface{}{
			"source":      resolved.Source,
			"resolved_at": resolved.ResolvedAt,
		}
	}
	return nil
}

File: src/runtime/environment.ts
/**
 * Environment Runtime
 * Provides type-safe access to resolved configuration
 */

import { Schema } from '../core/schema';
import { Resolver, ResolverOptions, ResolvedValue } from './resolver';

export interface EnvironmentOptions extends ResolverOptions {}

/**
 * Type-safe environment runtime
 */
export class Environment {
  private resolver: Resolver;
  private values: Record<string, unknown> = {};
  private resolved: ResolvedValue[] = [];
  private initialized = false;

  constructor(
    private schema: Schema,
    options: EnvironmentOptions = {}
  ) {
    this.resolver = new Resolver(schema, options);
  }

  /**
   * Initialize the environment
   */
  init(): void {
    const result = this.resolver.resolve();
    this.values = result.values;
    this.resolved = result.resolved;
    this.initialized = true;
  }

  /**
   * Get a value with type assertion
   */
  get<T = unknown>(name: string): T {
    this.ensureInitialized();
    if (!(name in this.values)) {
      throw new Error(`Variable "${name}" is not defined in schema`);
    }
    return this.values[name] as T;
  }

  /**
   * Get a value safely with optional return
   */
  getOptional<T = unknown>(name: string): T | undefined {
    this.ensureInitialized();
    return (this.values[name] as T | undefined) ?? undefined;
  }

  /**
   * Get a string value
   */
  getString(name: string): string {
    return this.get<string>(name);
  }

  /**
   * Get a number value
   */
  getNumber(name: string): number {
    return this.get<number>(name);
  }

  /**
   * Get an integer value
   */
  getInt(name: string): number {
    return this.get<number>(name);
  }

  /**
   * Get a boolean value
   */
  getBoolean(name: string): boolean {
    return this.get<boolean>(name);
  }

  /**
   * Get a JSON value
   */
  getJSON(name: string): unknown {
    return this.get(name);
  }

  /**
   * Check if a variable is secret
   */
  isSecret(name: string): boolean {
    const definition = this.schema.getVariable(name);
    return definition?.type.isSecret ?? false;
  }

  /**
   * Get resolution metadata for a variable
   */
  getMetadata(name: string) {
    const resolved = this.resolved.find((r) => r.name === name);
    return resolved?.metadata;
  }

  /**
   * Get all values (for debugging/inspection)
   */
  getAllValues(hideSecrets = true): Record<string, unknown> {
    this.ensureInitialized();
    if (!hideSecrets) {
      return { ...this.values };
    }

    const safe: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(this.values)) {
      if (this.isSecret(key)) {
        safe[key] = '[SECRET]';
      } else {
        safe[key] = value;
      }
    }
    return safe;
  }

  /**
   * Get all resolved values with metadata
   */
  getResolved(): ResolvedValue[] {
    this.ensureInitialized();
    return this.resolved.map((r) => ({
      ...r,
      value: this.isSecret(r.name) ? '[SECRET]' : r.value,
    }));
  }

  /**
   * Validate the environment
   */
  validate(): {
    valid: boolean;
    errors: Array<{ variable: string; error: string }>;
  } {
    this.ensureInitialized();
    const errors = this.resolved
      .filter((r) => !r.valid)
      .map((r) => ({
        variable: r.name,
        error: r.validationError || 'Unknown error',
      }));

    return {
      valid: errors.length === 0,
      errors,
    };
  }

  /**
   * Convert to record (for use with other libraries)
   */
  toObject(): Record<string, string> {
    this.ensureInitialized();
    const obj: Record<string, string> = {};
    for (const [key, value] of Object.entries(this.values)) {
      if (value !== null && value !== undefined) {
        obj[key] = String(value);
      }
    }
    return obj;
  }

  private ensureInitialized(): void {
    if (!this.initialized) {
      this.init();
    }
  }
}

/**
 * Create and initialize an environment
 */
export function createEnvironment(
  schema: Schema,
  options?: EnvironmentOptions
): Environment {
  const env = new Environment(schema, options);
  env.init();
  return env;
}
